packages/components/src/internal/popup-utils.ts

```diff
@@ -0,0 +1,173 @@
+  elementSize: Size;
+
+  /* Gap between popup and target elements. */
+  gap: number;
+
+  /* Preferred side of the target element to place the popup element on. */
+  preferredSide: Side;
+
+  /* Position of the target element. */
+  targetPosition: Position;
+
+  /* Size of the window. */
+  viewportSize: Size;
+}
+
+interface ComputePopupPositionResult {
+  position: Position;
+  side: Side;
+}
+
+/**
+ * Compute a popup's position based on its target element.
+ */
+export const computePopupPosition = (args: ComputePopupPositionArgs): ComputePopupPositionResult => {
+  const { elementSize, gap, preferredSide, targetPosition, viewportSize } = args;
+
+  const getPosition = (s: Side): Pick<Position, 'left' | 'top'> => {
+    const left = _.flow(
+      _.clamp(0, viewportSize.width - elementSize.width),
+      _.clamp(targetPosition.left - elementSize.width + 16, targetPosition.right - 16)
+    )((targetPosition.left + targetPosition.right) / 2 - elementSize.width / 2);
+
+    const top = _.flow(
+      _.clamp(0, viewportSize.height - elementSize.height),
+      _.clamp(targetPosition.top - elementSize.height + 16, targetPosition.bottom - 16)
+    )((targetPosition.top + targetPosition.bottom) / 2 - elementSize.height / 2);
+
+    switch (s) {
+      case 'top':
+        return { top: targetPosition.top - elementSize.height - gap, left };
+      case 'right':
+        return { left: targetPosition.right + gap, top };
+      case 'bottom':
+        return { top: targetPosition.bottom + gap, left };
+      case 'left':
+        return { left: targetPosition.left - elementSize.width - gap, top };
+      default:
+        throw new Error('Invalid side');
+    }
+  };
+
+  const position = getPosition(preferredSide);
+
+  // If the popup would overflow the viewport when placed on the preferred side of the target
+  // element, "flip" it to the other side of the element.
+  const maybeFlip = (d: Side): Side => {
+    switch (d) {
+      case 'top':
+        return position.top < 0 ? 'bottom' : 'top';
+      case 'right':
+        return position.left + elementSize.width >= viewportSize.width ? 'left' : 'right';
+      case 'bottom':
+        return position.top + elementSize.height >= viewportSize.height ? 'top' : 'bottom';
+      case 'left':
+        return position.left < 0 ? 'right' : 'left';
+      default:
+        throw new Error('Invalid side');
+    }
+  };
+
+  const finalSide = maybeFlip(preferredSide);
+
+  // Recompute position after side may have changed.
+  const finalPosition = getPosition(finalSide);
+
+  return {
+    position: {
+      ...finalPosition,
+      right: finalPosition.left + elementSize.width,
+      bottom: finalPosition.top + elementSize.height,
+    },
+    side: finalSide,
+  };
+};
+
+type BoundingRect = Size & Position;
+
+const toBoundingRect: (domRect: DOMRect) => BoundingRect = _.pick([
+  'width',
+  'height',
+  'top',
+  'right',
+  'bottom',
+  'left',
+]);
+
+type UseBoundingRectsSelector =
```

Comment:
generally speaking, I think we should export any types/interfaces that are used in exported functions, etc... makes things easier on the consumer side.

Reply:
Done in https://github.com/DataBiosphere/terra-ui/pull/4234/commits/ee07688c47955f1b581f1a9d5798a3079bbddc0b.